sound/soc/sof/sof-audio.h

```diff
@@ -473,6 +521,7 @@ struct snd_sof_dai {
 	int number_configs;
 	int current_config;
 	struct list_head list;	/* list in sdev dai list */
+	const struct snd_sof_dai_ops *ops;
```

Comment:
I would argue that this is Intel only and can be simply managed internally within sound/soc/sof/intel just fine by using an array for the second level of callbacks.

In essence the change would be similar but w/o touching core sof files.

Reply:
@ujfalusi I would have made this Intel-only too but look at chained-dma mode. It applies to individual pipelines and both the blanket topology like the dsp-less mode. So I need the ops to be different for the chained-dma DAI widget and the regular pipeline-based DAI widget.

Reply:
I still believe this is a Intel only thing and there should not be changes in core for this.

If the argument/reason for this is to pave the way for the ChainDMA support, then I still think this is not a correct way.

The ChainDMA for the stream/path/dai is coming from the topology, so using a flag in core components to say that this DAI should be used with ChainDMA is OK.

In Intel code the `hda_dai_get_ops()` can still look up the dai widget, check the flag and decide which set of callback ops it will return.
This series then would not need to touch the core and focus on what's important here: re-arrange the Intel specific HDA code for new features.

I still argue that the core should not know anything about this peculiarity.